File: src/main/java/com/dude/dms/ui/builder/rules/RuleBuilderFactory.java
package com.dude.dms.ui.builder.rules;

import com.dude.dms.backend.brain.parsing.PlainTextRuleValidator;
import com.dude.dms.backend.brain.parsing.RegexRuleValidator;
import com.dude.dms.backend.data.Tag;
import com.dude.dms.backend.data.docs.Doc;
import com.dude.dms.backend.data.rules.PlainTextRule;
import com.dude.dms.backend.data.rules.RegexRule;
import com.dude.dms.backend.service.DocService;
import com.dude.dms.backend.service.PlainTextRuleService;
import com.dude.dms.backend.service.RegexRuleService;
import com.dude.dms.backend.service.TagService;
import com.dude.dms.ui.builder.BuilderFactory;
import com.dude.dms.ui.builder.Factory;

import java.util.Map;
import java.util.Set;

public class RuleBuilderFactory extends Factory {

    private final TagService tagService;
    private final PlainTextRuleValidator plainTextRuleValidator;
    private final RegexRuleValidator regexRuleValidator;
    private final PlainTextRuleService plainTextRuleService;
    private final RegexRuleService regexRuleService;
    private final DocService docService;

    public RuleBuilderFactory(BuilderFactory builderFactory, TagService tagService, PlainTextRuleValidator plainTextRuleValidator,
                              RegexRuleValidator regexRuleValidator, PlainTextRuleService plainTextRuleService, RegexRuleService regexRuleService,
                              DocService docService) {
        super(builderFactory);
        this.tagService = tagService;
        this.plainTextRuleValidator = plainTextRuleValidator;
        this.regexRuleValidator = regexRuleValidator;
        this.plainTextRuleService = plainTextRuleService;
        this.regexRuleService = regexRuleService;
        this.docService = docService;
    }

    public PlainTextRuleCardBuilder plainTextCard(PlainTextRule rule) {
        return new PlainTextRuleCardBuilder(builderFactory, rule, plainTextRuleValidator, tagService);
    }

    public RegexRuleCardBuilder regexCard(RegexRule rule) {
        return new RegexRuleCardBuilder(builderFactory, rule, regexRuleValidator, tagService);
    }

    public PlainTextRuleDialogBuilder plainDialog() {
        return new PlainTextRuleDialogBuilder(tagService, plainTextRuleService);
    }

    public RegexRuleDialogBuilder regexDialog() {
        return new RegexRuleDialogBuilder(tagService, regexRuleService);
    }

    public RuleRunnerDialogBuilder ruleRunnerDialog(Map<Doc, Set<Tag>> result) {
        return new RuleRunnerDialogBuilder(result, docService);
    }
}

File: src/main/java/com/dude/dms/ui/builder/rules/RuleBuilderFactory.kt
package com.dude.dms.ui.builder.rules

import com.dude.dms.brain.parsing.PlainTextRuleValidator
import com.dude.dms.brain.parsing.RegexRuleValidator
import com.dude.dms.backend.data.Tag
import com.dude.dms.backend.data.docs.Doc
import com.dude.dms.backend.data.mails.MailFilter
import com.dude.dms.backend.data.rules.PlainTextRule
import com.dude.dms.backend.data.rules.RegexRule
import com.dude.dms.backend.service.*
import com.dude.dms.brain.events.EventManager
import com.dude.dms.brain.mail.MailManager
import com.dude.dms.ui.builder.BuilderFactory
import com.dude.dms.ui.builder.Factory

class RuleBuilderFactory(
        builderFactory: BuilderFactory,
        private val tagService: TagService,
        private val plainTextRuleValidator: PlainTextRuleValidator,
        private val regexRuleValidator: RegexRuleValidator,
        private val plainTextRuleService: PlainTextRuleService,
        private val regexRuleService: RegexRuleService,
        private val docService: DocService,
        private val mailFilterService: MailFilterService,
        private val mailManager: MailManager,
        private val eventManager: EventManager
) : Factory(builderFactory) {

    fun plainTextCard(rule: PlainTextRule)= PlainTextRuleCardBuilder(builderFactory, rule, plainTextRuleValidator, tagService, eventManager)

    fun regexCard(rule: RegexRule) = RegexRuleCardBuilder(builderFactory, rule, regexRuleValidator, tagService, eventManager)

    fun mailCard(mailFilter: MailFilter) = MailFilterCardBuilder(builderFactory, mailFilter)

    fun plainCreateDialog() = PlainTextRuleCreateDialogBuilder(builderFactory, plainTextRuleService)

    fun regexCreateDialog() = RegexRuleCreateDialogBuilder(builderFactory, regexRuleService)

    fun mailCreateDialog() = MailFilterCreateDialogBuilder(mailFilterService, mailManager)

    fun plainEditDialog(rule: PlainTextRule)= PlainTextRuleEditDialogBuilder(builderFactory, rule, plainTextRuleService)

    fun regexEditDialog(rule: RegexRule) = RegexRuleEditDialogBuilder(builderFactory, rule, regexRuleService)

    fun mailEditDialog(mailFilter: MailFilter) = MailFilterEditDialogBuilder(mailFilter, mailFilterService, mailManager)

    fun ruleRunnerDialog(result: Map<Doc, Set<Tag>>) = RuleRunnerDialogBuilder(result, docService)
}